线程组中不可控异常的处理
=========================================================
我们学习了如何在线程对象中处理非捕获异常，另一种可行的做法是，**建立一个方法来捕获线程组中的任何线程对象
抛出的非捕获异常**。

示例：
```java
import java.util.Random;

public class Main3 {

    public static void main(String[] args) {
        MyThreadGroup threadGroup = new MyThreadGroup("MyThreadGroup");
        Task task = new Task();
        for (int i = 0; i < 2; i++) {
            Thread t = new Thread(threadGroup, task);
            t.start();
        }
    }

    private static class MyThreadGroup extends ThreadGroup {

        //ThreadGroup没有默认构造器
        public MyThreadGroup(String name) {
            super(name);
        }

        //重写这个方法，当线程组中的任何线程对象抛出异常的时候，这个方法就会被调用
        @Override
        public void uncaughtException(Thread t, Throwable e) {
            System.out.printf("The thread %s has thrown an Exception\n", t.getId());
            e.printStackTrace(System.out);
            System.out.printf("Terminating the rest of the Threads\n");
            interrupt();
        }
    }

    private static class Task implements Runnable {

        @Override
        public void run() {
            int result;
            Random random = new Random(Thread.currentThread().getId());
            while (true) {
                result = 1000 / ((int) (random.nextDouble() * 1000));
                System.out.printf("%s : %d\n", Thread.currentThread().getId(), result);
                if (Thread.currentThread().isInterrupted()) {
                    System.out.printf("%d : Interrupted\n", Thread.currentThread().getId());
                    return;
                }
            }
        }
    }
}
```
